pandas/core/index.py

```diff
@@ -2511,6 +2521,206 @@ def invalid_op(self, other=None):
 Index._add_numeric_methods_disabled()
 Index._add_logical_methods()
 
+class CategoricalIndex(Index):
+    """
+
+    Immutable Index implementing an ordered, sliceable set. CategoricalIndex
+    represents a sparsely populated Index with an underlying Categorical.
+
+    Parameters
+    ----------
+    data : array-like (1-dimensional)
+    categories : optional categories for the CategoricalIndex
+    copy : bool
+        Make a copy of input ndarray
+    name : object
+        Name to be stored in the index
+
+    """
+
+    _typ = 'categoricalindex'
+    _engine_type = _index.Int64Engine
+    _attributes = ['name','categories']
+
+    def __new__(cls, data=None, categories=None, dtype=None, copy=False, name=None, fastpath=False, ordered=None, **kwargs):
+
+        def create_categorical(data=data, categories=categories):
+            if categories is not None:
+                data = data.set_categories(categories)
+            if not data.ordered:
+                data = data.as_ordered()
+            return data
+
+        if fastpath:
+            return cls._simple_new(data, name=name)
+
+        if ordered is not None:
+            raise ValueError("CategoricalIndex are by definition ordered")
+
+        if isinstance(data, com.ABCCategorical):
+            data = create_categorical(data, categories)
+        elif data is None or np.isscalar(data):
+            cls._scalar_data_error(data)
+        elif isinstance(data, CategoricalIndex):
+            data = data._data
+            data = create_categorical(data, categories)
+        else:
+            from pandas.core.categorical import Categorical
+            data = Categorical(data, categories=categories, ordered=True)
+
+        if copy:
+            data = data.copy()
+
+        return cls._simple_new(data, name=name)
+
+    @classmethod
+    def _simple_new(cls, values, name=None, categories=None, **kwargs):
+        result = object.__new__(cls)
+
+        if not isinstance(values, com.ABCCategorical):
+            from pandas.core.categorical import Categorical
+            values = Categorical(values, categories=categories, ordered=True)
+        elif categories is not None:
+            values = values.set_categories(categories)
+
+        result._data = values
+        result.name = name
+        for k, v in compat.iteritems(kwargs):
+            setattr(result,k,v)
+
+        result._reset_identity()
+        return result
+
+    def equals(self, other):
```

Comment:
Does `CategoricalIndex(['a', 'b'], categories=['a', 'b']) == CategoricalIndex(['a', 'b'], categories=['a', 'b', 'c'])` return True?.

I just checked on Categoricals and we raise a `TypeError` if the categories aren't identical.
```python
In [1]: c1 = pd.Categorical(['a', 'b'], categories=['a', 'b'])

In [2]: c2 = pd.Categorical(['a', 'b'], categories=['a', 'b', 'c'])

In [5]: c1 == c2
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-5-d8d43a43a02a> in <module>()
----> 1 c1 == c2

/Users/tom.augspurger/Envs/py3/lib/python3.4/site-packages/pandas-0.16.0_19_g8d2818e-py3.4-macosx-10.10-x86_64.egg/pandas/core/categorical.py in f(self, other)
     38             if (len(self.categories) != len(other.categories)) or \
     39                     not ((self.categories == other.categories).all()):
---> 40                 raise TypeError("Categoricals can only be compared if 'categories' are the same")
     41             if not (self.ordered == other.ordered):
     42                 raise TypeError("Categoricals can only be compared if 'ordered' is the same")

TypeError: Categoricals can only be compared if 'categories' are the same
```